src/feature/hs/hs_intropoint.c

```diff
@@ -181,6 +181,108 @@ hs_intro_send_intro_established_cell,(or_circuit_t *circ))
   return ret;
 }
 
+static void
+handle_establish_intro_cell_dos_extension(
+                                const trn_cell_extension_field_t *field,
+                                or_circuit_t *circ)
+{
+  ssize_t ret;
+  uint64_t intro2_rate_per_sec = 0, intro2_burst_per_sec = 0;
+  trn_cell_extension_dos_t *dos = NULL;
+
+  tor_assert(field);
+  tor_assert(circ);
+
+  ret = trn_cell_extension_dos_parse(&dos,
+                 trn_cell_extension_field_getconstarray_field(field),
+                 trn_cell_extension_field_getlen_field(field));
+  if (ret < 0) {
+    goto end;
+  }
+
+  for (size_t i = 0; i < trn_cell_extension_dos_get_n_params(dos); i++) {
+    const trn_cell_extension_dos_param_t *param =
+      trn_cell_extension_dos_getconst_params(dos, i);
+    if (BUG(param == NULL)) {
+      goto end;
+    }
+
+    switch (trn_cell_extension_dos_param_get_type(param)) {
+    case TRUNNEL_DOS_PARAM_TYPE_INTRO2_RATE_PER_SEC:
+      intro2_rate_per_sec = trn_cell_extension_dos_param_get_value(param);
+      break;
+    case TRUNNEL_DOS_PARAM_TYPE_INTRO2_BURST_PER_SEC:
+      intro2_burst_per_sec = trn_cell_extension_dos_param_get_value(param);
+      break;
+    default:
+      goto end;
+    }
+  }
+
+  /* Validation. A value of 0 on either of them means the defenses are
```

Comment:
It would be really nice if this value validation was the same code that does the torrc validation, to minimize the desynch that can happen between torrc and intro point validation here (so that we only have possible desynch between tor versions). Do you think this is possible?

Reply:
Hmm... possibly if I modularize the check in `hs_config.c` and expose the functions publicly so `hs_intropoint.c` could go for it.

Not a bad idea at all!

Reply:
Oh wait now I remember why it is separated... It is because with the torrc options, we parse strings and use helper functions for that that do check for boundaries:

See `helper_parse_uint64()` ... so we pass the default values to that function and we get a boundary checks.

I'll at least use the `#define` that exists for the default values of the DoS defenses in `hs_intropoint.c`.

Reply:
ACK!